packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -271,17 +271,107 @@ class _UntilTextBoundary extends TextBoundary {
   }
 }
 
+class _TextLayout {
+  _TextLayout._(this._paragraph);
+
+  // This field is not final because the owner TextPainter could create a new
+  // ui.Paragraph with the exact same text layout (for example, when only the
+  // color of the text is changed).
+  ui.Paragraph _paragraph;
+
+  // Unfortunately, using full precision floating point here causes bad layouts
+  // because floating point math isn't associative. If we add and subtract
+  // padding, for example, we'll get different values when we estimate sizes and
+  // when we actually compute layout because the operations will end up associated
+  // differently. To work around this problem for now, we round fractional pixel
+  // values up to the nearest whole pixel value. The right long-term fix is to do
+  // layout using fixed precision arithmetic.
+  static double _applyFloatingPointHack(double layoutValue) => layoutValue.ceilToDouble();
```

Comment:
we really should stop doing this

Reply:
also the comment is just wrong (it's been wrong for a long time). we looked at doing what it says, but we decided not to, many years ago. if we're going to make a breaking change anyway, we might as well remove this at the same time...

Reply:
based on the conversation in https://github.com/flutter/flutter/issues/31707, it doesn't seem to be safe to remove this hack yet, since widths/maxIntrinsicWidths are reported in floats and the input widths are doubles

Reply:
After removing the rounding in the framework and skparagraph, and tuning down the harfbuzz scale from 1^16 to 1^10, with this change there are only 50 framework failures and nothing seems to be overflowing. So I guess that's doable but I'd like to try to do that in a separate PR.

Reply:
@Hixie still this breaks 500+ scuba tests. But I'm not sure how I can write a migration guide for this one.